nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/marshallers/json/enrichers/AbstractJsonEnricher.java

```diff
@@ -66,19 +68,20 @@ public void write(Enriched<EntityType> enrichable, JsonGenerator jg) {
             // Write to a temporary output in case of exception during write()
             tb.writeStartObject();
             write(tb, enrichable.getEntity());
+            tb.writeEndObject();
             tb.flush();
             // Add the complete, well-formed content to the real output
             try (JsonParser parser = tb.asParser()) {
                 parser.nextToken(); // ignoring START_OBJECT
-                while (parser.nextToken() == JsonToken.FIELD_NAME) {
+                while (parser.nextToken() == FIELD_NAME) {
                     jg.copyCurrentStructure(parser);
                 }
-                if (parser.currentToken() != null) {
-                    log.error("Enricher {} returned invalid output {}", name, tb.toString());
+                if (parser.currentToken() != END_OBJECT) {
+                    log.error("Enricher: {} failed on current token: {}, output: {}", name, parser.currentToken(), tb);
```

Comment:
I'm wondering... what does log error `tb` produce? If it's what we currently see in the logs (see below), I don't think it's interesting as the output is truncated anyway.

```
[TokenBuffer: START_OBJECT, FIELD_NAME(acls), START_ARRAY, START_OBJECT, FIELD_NAME(name), VALUE_STRING, FIELD_NAME(aces), START_ARRAY, START_OBJECT, FIELD_NAME(id), VALUE_STRING, FIELD_NAME(username), START_OBJECT, FIELD_NAME(entity-type), ...
```

Reply:
Yes this is the output of tb which can be verbose for not much value indeed.
(The case in which it can help is if the output is not long enough to be truncated so we can see what other enrichment was already written before)
I let you decide!

Reply:
Now we can see the temporary json content we were trying to write
```
if (parser.currentToken() != END_OBJECT) {
    log.error("Enricher: {} failed on current token: {}, output to write: {}", name,
    parser.currentToken(), MAPPER.readTree(tb.asParser()));
}
```

Reply:
As it may be costly (?) what about logging the whole JSON in `debug` only? WDYT @kevinleturc @efge?
Anyway, you should use method references/lambda when logging "costly" stuff:
```
() -> MAPPER.readTree(tb.asParser())
```

Reply:
readTree is not a functional interface.
If I wrap the call in a Supplier i need to try/catch the exceptions it throws:
```
if (parser.currentToken() != END_OBJECT) {
    Supplier<String> s = () -> {
        try {
            return MAPPER.readTree(tb.asParser());
        } catch (IOException e) {
            log.error("Failed to read content, Enricher: {}", name, e);
            return "";
        }
    };
    log.error("Enricher: {} failed on current token: {}, output to write: {}", name,
            parser.currentToken(), s.get());
}
```

Reply:
Since all this is happening in case of error, the condition of failing is acting as a guard.
Which means the expensive logging will only happen in case it is needed and the code would stay simple.

Reply:
Yes it's ok to have a slow path in error cases. Make the code simple.

Reply:
I agree with @troger, as it's a valide usage we may have a lot of logs such as it. Apart from Java performance issue, we could have a lot of long logs, which may have an impact on logs aggregator.

Furthermore, and also as a separate topic, I would try/catch the readTree call as we're handling invalid stuff this call can crash too.

Reply:
@kevinleturc you're saying we expect to have a lot of exceptions thrown by enrichers? Or enrichers emitting invalid JSON? Why?
+1 on the readTree protection

Reply:
> @kevinleturc you're saying we expect to have a lot of exceptions thrown by enrichers? Or enrichers emitting invalid JSON? Why?

After re-checking code it's indeed a specific code block that may not be executed in all failure cases and especially not in the case we want to cover.

Reply:
> Furthermore, and also as a separate topic, I would try/catch the readTree call as we're handling invalid stuff this call can crash too.

It is catched in the main catch appended to the try resource. Or do you mean a sub catch close just at the logging instruction to be more specific in error handling ?

Reply:
I meant a sub catch indeed.

Reply:
OK then to avoid nesting a try catch again (sonar unhappy), I first made a protected method to readTree within a try/catch. 
But then It felt like even more cumbersome than going with the Supplier. So I went with the Supplier which allows the performance gain.